### DISCUSSION TITLE
bindEvents for hits not available?

### DISCUSSION THREAD

**Discussion by NiklasBr** [algolia/instantsearch#5447] [Q&A] [answered]
I'm having trouble implementing the Insights Clients analytics section for Instantsearch. The short version (pagination and config removed)  looks like this:

```js
import { index, hits, } from "instantsearch.js/es/widgets";

export const pageHits = () =>
	index({
		indexName: window.algoliaIndexPrefix + window.algoliaSite + "_page",
	}).addWidgets([
		hits({
			container: "#pages",
			templates: {
				item: (hit, { html, components }) => html `
					<a class="pages" href="${hit.url}"
					data-insights-object-id="${hit.objectID}"
					data-insights-position="${hit.__position}"
					data-insights-query-id="${hit.__queryID}">
						${components.Highlight({ attribute: 'title', hit })}
				</a>
			`,
			},
		}),
	]);
```

The above works, but when adding the `bindEvent` param to `item` [which should work according to this documentation](https://www.algolia.com/doc/guides/building-search-ui/upgrade-guides/js/?client=js#insightsclient-at-instantsearch) page, we get errors:

```js
		hits({
			container: "#pages",
			templates: {
				item: (hit, bindEvent, { html, components }) => html `
					<a class="pages" href="${hit.url}"
					data-insights-object-id="${hit.objectID}"
					data-insights-position="${hit.__position}"
					data-insights-query-id="${hit.__queryID}"
					${bindEvent('click', hit, 'Page clicked')}>
						${components.Highlight({ attribute: 'title', hit })}
				</a>
			`,
			},
		}),
```

What's wrong here?

**Comment by Haroenv** [comment, marked as answer]
BindEvent does not work with html templates, as it uses the interpolations to define what is an attribute, element etc. Instead, you can use sendEvent:

```js
hits({
  // ...
  templates: {
    item(hit, { html, components, sendEvent }) {
      return html`
        <h2>${components.Highlight({ hit, attribute: 'name' })}</h2>
        <p>${components.Snippet({ hit, attribute: 'description' })}</p>
        <button
          onClick="${() => {
            sendEvent('conversion', hit, 'Product Added');
          }}"
        >
          Add to cart
        </button>
      `;
    },
  },
});
```

I'll update the docs to discourage using bindEvent for that reason :)

**Reply by NiklasBr** [reply]
Thanks for both the correction and the quick reply!

Do we need to add anything to the middleWare? We tried to implement [like the documentation here indicates](https://www.algolia.com/doc/guides/building-search-ui/upgrade-guides/js/?client=js#analytics-widget).

```js
const insightsMiddleware = createInsightsMiddleware({
    insightsClient: window.aa,
    insightsInitParams: {
        userHasOptedOut: !window._ptg.dataProviders.analyticsAllowed(),
        userToken: window._ptg.dataProviders.getVisitorId(),
    },
    onEvent: (event, aa) => {
        const { insightsMethod, payload, widgetType, eventType } = event;
        aa(insightsMethod, payload);
    },
});
```

*edit*

If I `console.log({event, aa});` in `onEvent` above there is a difference between the `click` or `conversion` event and the regular `view` event. Is this intended?

View event:
![image](https://user-images.githubusercontent.com/279826/214616859-2e42c959-daa4-4597-80e2-10a5bbfde8f0.png)

sendEvent:
![image](https://user-images.githubusercontent.com/279826/214617207-02f530bc-46d1-49a2-844c-103a9a3cf50c.png)

Notice how the event itself is just a string in the latter instance, but a full Object in the former.

**Reply by Haroenv** [reply]
I'm sorry, I just found a bug due to your report. There were two `sendEvent` functions, one transforming the arguments into a complete object (like bindEvent does already), and one passing the first argument through immediately. We were using the wrong function in the tagged template.

I've fixed this in https://github.com/algolia/instantsearch/pull/5457, so you can either wait until that's merged, or use `bindEvent` without the html tagged template for now. You also could pass the full object with insightsMethod etc. to the first argument, which will still work after the PR as well :)

**Reply by NiklasBr** [reply]
Thanks again! Will await the next release

**Reply by Haroenv** [reply]
the latest release has happened, so you can use 4.50.1!

**Reply by NiklasBr** [reply]
I only looked at the *Releases* and not at *Tags* and keep forgetting that GitHub keeps them very separate 👍

**Reply by Haroenv** [reply]
Yes, we recently migrated our releases tool and didn't yet make sure every tag gets a release, good to know it would help you when we fix that

**Reply by NiklasBr** [reply]
![image](https://user-images.githubusercontent.com/279826/215053932-16b1d9dc-a59d-4fff-9078-b34cd546b0d4.png)
